File: androidRandomAssigner/src/main/java/com/barryalan/randomassignerkmm/android/presentation/sidework_list/AppViewModel.kt
package com.barryalan.randomassignerkmm.android.presentation.sidework_list

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barryalan.randomassignerkmm.domain.model.*
import com.barryalan.randomassignerkmm.domain.util.GenericMessageInfoQueueUtil
import com.barryalan.randomassignerkmm.domain.util.Queue
import com.barryalan.randomassignerkmm.interactors.employee_edit.DeleteEmployee
import com.barryalan.randomassignerkmm.interactors.employee_edit.EditEmployee
import com.barryalan.randomassignerkmm.interactors.employee_edit.GetEmployees
import com.barryalan.randomassignerkmm.interactors.sidework_edit.DeleteSidework
import com.barryalan.randomassignerkmm.interactors.sidework_edit.EditSidework
import com.barryalan.randomassignerkmm.interactors.sidework_edit.GetSideworks
import com.barryalan.randomassignerkmm.interactors.sidework_list.*
import com.barryalan.randomassignerkmm.presentation.AppEvents
import com.barryalan.randomassignerkmm.presentation.AppState
import dagger.hilt.android.lifecycle.HiltViewModel
import java.lang.Exception
import java.util.*
import javax.inject.Inject

@HiltViewModel
class AppViewModel
@Inject
constructor(
    private val shuffleSideworks: ShuffleSideworks,
    private val getEmployees: GetEmployees,
    private val getSideworks: GetSideworks,
    private val deleteEmployee: DeleteEmployee,
    private val deleteSidework: DeleteSidework,
    private val editEmployee: EditEmployee,
    private val editSidework: EditSidework
) : ViewModel() {

    val state: MutableState<AppState> = mutableStateOf(AppState())

    fun onTriggerEvent(event: AppEvents) {
        when (event) {

            is AppEvents.OnRemoveHeadMessageFromQueue -> removeHeadMessage()

            AppEvents.ShuffleSideworks -> assignSideworks()

            AppEvents.ToggleEditSideworks -> toggleEditSideworks()
            AppEvents.ToggleEditEmployees -> toggleEditEmployees()

            is AppEvents.ToggleEditSidework -> toggleEditSidework(event.sidework)
            is AppEvents.ToggleEditEmployee -> toggleEditEmployee(event.employee)

            is AppEvents.ToggleTodoToday -> saveSidework(event.sidework)
            is AppEvents.ToggleIsHere -> saveEmployee(event.employee)

            is AppEvents.SetNewSideworkName -> setNewSideworkName(event.name)
            is AppEvents.SetNewEmployeeName -> setNewEmployeeName(event.name)

            is AppEvents.DeleteSidework -> deleteSidework(event.sideworkID)
            is AppEvents.DeleteEmployee -> deleteEmployee(event.employeeID)

            is AppEvents.SaveSidework -> saveSidework(event.sidework)
            is AppEvents.SaveEmployee -> saveEmployee(event.employee)

            else -> appendToMessageQueue(
                GenericMessageInfo.Builder()
                    .id(UUID.randomUUID().toString())
                    .title("Error")
                    .uiComponentType(UIComponentType.Dialog)
                    .description("Invalid event")
            )
        }
    }

    private fun removeHeadMessage() {
        try {
            val queue = state.value.queue
            queue.remove()
            state.value = state.value.copy(queue = Queue(mutableListOf())) //force recompose
            state.value = state.value.copy(queue = queue)
        } catch (e: Exception) {
            //nothing to remove, queue is empty
        }
    }

    private fun assignSideworks() {
        shuffleSideworks.execute().collectCommon(viewModelScope) { dataState ->

            dataState.data?.let { sideworks ->
                state.value = state.value.copy(shuffledSideworks = sideworks)
            }

            dataState.message?.let { message ->
                appendToMessageQueue(message)
            }

        }
    }

    private fun toggleEditSideworks() {
        getSideworks.execute().collectCommon(viewModelScope) { dataState ->

            dataState.data?.let { sideworks ->
                state.value = state.value.copy(
                    sideworks = sideworks,
                    isSideworksEditShowing = !state.value.isSideworksEditShowing
                )
            }

            dataState.message?.let { message ->
                appendToMessageQueue(message)
            }


        }
    }

    private fun toggleEditEmployees() {
        getEmployees.execute().collectCommon(viewModelScope) { dataState ->

            dataState.data?.let { employees ->
                state.value = state.value.copy(
                    employees = employees,
                    isEmployeesEditShowing = !state.value.isEmployeesEditShowing
                )
            }

            dataState.message?.let { message ->
                appendToMessageQueue(message)
            }


        }

    }

    private fun toggleEditSidework(sidework: Sidework) {
        if (state.value.selectedSideworkID == sidework.id) {
            state.value = state.value.copy(
                newSideworkName = "",
                selectedSideworkID = "",
                isSideworkEditShowing = false,
                sideworkButtonText = "Add"
            )
        } else {
            state.value = state.value.copy(
                newSideworkName = sidework.name,
                selectedSideworkID = sidework.id,
                isSideworkEditShowing = true,
                sideworkButtonText = "Save"
            )
        }
    }

    private fun toggleEditEmployee(employee: Employee) {
        if (state.value.selectedEmployeeID == employee.id) {
            state.value = state.value.copy(
                newEmployeeName = "",
                selectedEmployeeID = "",
                isEmployeeEditShowing = false,
                employeeButtonText = "Add"
            )
        } else {
            state.value = state.value.copy(
                newEmployeeName = employee.name,
                selectedEmployeeID = employee.id,
                isEmployeeEditShowing = true,
                employeeButtonText = "Save"
            )
        }
    }

    private fun setNewSideworkName(name: String) {
        state.value = state.value.copy(newSideworkName = name)
    }

    private fun setNewEmployeeName(name: String) {
        state.value = state.value.copy(newEmployeeName = name)
    }

    private fun deleteSidework(sideworkID: String) {
        deleteSidework.execute(sideworkID).collectCommon(viewModelScope) { dataState ->

            dataState.data?.let { sideworks ->
                state.value = state.value.copy(sideworks = sideworks)
            }

            dataState.message?.let { message ->
                appendToMessageQueue(message)
            }

        }
    }

    private fun deleteEmployee(employeeID: String) {
        deleteEmployee.execute(employeeID).collectCommon(viewModelScope) { dataState ->

            dataState.data?.let { employees ->
                state.value = state.value.copy(employees = employees)
            }

            dataState.message?.let { message ->
                appendToMessageQueue(message)
            }

        }
    }

    private fun saveSidework(sidework: Sidework) {
        editSidework.execute(sidework).collectCommon(viewModelScope) { dataState ->

            dataState.data?.let { sideworks ->
                state.value = state.value.copy(
                    sideworks = sideworks,
                    newSideworkName = "",
                    sideworkButtonText = "Add",
                    selectedSideworkID = "",
                    isSideworkEditShowing = false
                )
            }

            dataState.message?.let { message ->
                appendToMessageQueue(message)
            }

        }
    }

    private fun saveEmployee(employee: Employee) {
        editEmployee.execute(employee).collectCommon(viewModelScope) { dataState ->

            dataState.data?.let { employees ->

                state.value = state.value.copy(
                    employees = employees,
                    newEmployeeName = "",
                    employeeButtonText = "Add",
                    selectedEmployeeID = "",
                    isEmployeeEditShowing = false
                )
            }

            dataState.message?.let { message ->
                appendToMessageQueue(message)
            }

        }
    }

    private fun appendToMessageQueue(messageInfo: GenericMessageInfo.Builder) {
        if (!GenericMessageInfoQueueUtil()
                .doesMessageAlreadyExistInQueue(
                    queue = state.value.queue, messageInfo = messageInfo.build()
                )
        ) {
            val queue = state.value.queue
            queue.add(messageInfo.build())
            state.value = state.value.copy(queue = Queue(mutableListOf()))
            state.value = state.value.copy(queue = queue)
        }

    }
}

File: iosRandomAssigner/iosRandomAssigner/presentation/main_screen/AppViewModel.swift
//
//  AppViewModel.swift
//  iosRandomAssigner
//
//  Created by Alan Ramirez on 10/27/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import SwiftUI
import shared


class AppViewModel:ObservableObject {
    
    //MARK: - VARIABLES (Use cases)
    let shuffleSideworksUseCase:ShuffleSideworks
    let getEmployeesUseCase:GetEmployees
    let getSideworksUseCase:GetSideworks
    let deleteEmployeeUseCase:DeleteEmployee
    let deleteSideworkUseCase:DeleteSidework
    let editEmployeeUseCase:EditEmployee
    let editSideworkUseCase:EditSidework
    private let cacheModule:CacheModule
    
    @Published var state: AppState = AppState()
    
    @Published var showDialog:Bool = false
    
    //MARK: - INIT
    init(){
        self.cacheModule = CacheModule()
        
        self.shuffleSideworksUseCase = ShuffleSideworksModule(cacheModule: self.cacheModule).shuffleSideworks
        self.getEmployeesUseCase = GetEmployeesModule(cacheModule: self.cacheModule).getEmployees
        self.getSideworksUseCase = GetSideworksModule(cacheModule: self.cacheModule).getSideworks
        self.deleteEmployeeUseCase = DeleteEmployeeModule(cacheModule: self.cacheModule).deleteEmployee
        self.deleteSideworkUseCase = DeleteSideworkModule(cacheModule: self.cacheModule).deleteSidework
        self.editEmployeeUseCase = EditEmployeeModule(cacheModule: self.cacheModule).editEmployee
        self.editSideworkUseCase = EditSideworkModule(cacheModule: self.cacheModule).editSidework
        
    }
    
    //MARK: - UPDATE STATE
    func updateState(
        queue:Queue<GenericMessageInfo>? = nil
    ){
        let currentState:AppState  = (self.state.copy() as! AppState)
        self.state = self.state.doCopy(
            shuffledSideworks: currentState.shuffledSideworks,
            sideworks:  currentState.sideworks,
            employees: currentState.employees,
            isSideworksEditShowing: currentState.isSideworksEditShowing,
            isEmployeesEditShowing: currentState.isEmployeesEditShowing,
            isSideworkEditShowing: currentState.isSideworksEditShowing,
            isEmployeeEditShowing: currentState.isEmployeeEditShowing,
            newSideworkName: currentState.newSideworkName,
            newEmployeeName: currentState.newEmployeeName,
            selectedSideworkID: currentState.selectedSideworkID,
            selectedEmployeeID: currentState.selectedEmployeeID,
            employeeButtonText: currentState.employeeButtonText,
            sideworkButtonText: currentState.sideworkButtonText,
            queue: queue ?? currentState.queue
        )
        shouldShowDialog()
    }
    
    //MARK: - ON TRIGGER EVENT
    func onTriggerEvent(stateEvent:AppEvents){
        switch stateEvent {
            
        case is AppEvents.ShuffleSideworks:
            assignSideworks()
            
        case is AppEvents.DeleteEmployee:
            deleteEmployee(employeeID:(stateEvent as! AppEvents.DeleteEmployee).employeeID)
            
        case is AppEvents.DeleteSidework:
            deleteSidework(sideworkID:(stateEvent as! AppEvents.DeleteSidework).sideworkID)
            
        case is AppEvents.SaveEmployee:
            saveEmployee(employee:(stateEvent as! AppEvents.SaveEmployee).employee)
            
        case is AppEvents.SaveSidework:
            saveSidework(sidework:(stateEvent as! AppEvents.SaveSidework).sidework)
            
        case is AppEvents.SetNewEmployeeName:
            setNewEmployeeName(name:(stateEvent as! AppEvents.SetNewEmployeeName).name)
            
        case is AppEvents.SetNewSideworkName:
            setNewSideworkName(name:(stateEvent as! AppEvents.SetNewSideworkName).name)
            
        case is AppEvents.ToggleEditEmployee:
            toggleEditEmployee(employee:(stateEvent as! AppEvents.ToggleEditEmployee).employee)
            
        case is AppEvents.ToggleEditSidework:
            toggleEditSidework(sidework:(stateEvent as! AppEvents.ToggleEditSidework).sidework)
            
        case is AppEvents.ToggleEditEmployees:
            toggleEditEmployees()
            
        case is AppEvents.ToggleEditSideworks:
            toggleEditSideworks()
            
        case is AppEvents.ToggleIsHere:
            saveEmployee(employee:(stateEvent as! AppEvents.ToggleIsHere).employee)
            
        case is AppEvents.ToggleTodoToday:
            saveSidework(sidework:(stateEvent as! AppEvents.ToggleTodoToday).sidework)
            
        case is AppEvents.OnRemoveHeadMessageFromQueue:
            removeHeadFromQueue()
            
        default:
            doNothing()
        }
    }
    
    //MARK: - TOGGLE EDIT SIDEWORKS
    private func toggleEditSideworks(){
        self.showEditSideworks()
        
        getSideworksUseCase.execute()
            .collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                    if dataState != nil {
                        let data = dataState?.data
                        let message = dataState?.message
                        
                        
                        self.updateState()
                        
                        if(data != nil){
                            self.setSideworks(sideworks: data as! [Sidework])
                        }
                        
                        if message != nil{
                            self.handleMessageByUIComponentType(message!.build())
                        }
                    }
                })
    }
    
    //MARK: - TOGGLE EDIT EMPLOYEES
    private func toggleEditEmployees(){
        
        self.showEditEmployees()
        
        getEmployeesUseCase.execute()
            .collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                    if dataState != nil {
                        let data = dataState?.data
                        let message = dataState?.message
                        
                        
                        self.updateState()
                        
                        if(data != nil){
                            if (data as! [Sidework]).isEmpty {
                                print("toggleEditSideworks, empty")
                            }
                            self.setEmployees(employees: data as! [Employee])
                        }
                        
                        if message != nil{
                            self.handleMessageByUIComponentType(message!.build())
                        }
                    }
                })
    }
    
    //MARK: - TOGGLE EDIT SIDEWORK
    private func toggleEditSidework(sidework:Sidework){
        var currentState = (self.state.copy() as! AppState)
        if(state.selectedSideworkID == sidework.id){
            self.state = self.state.doCopy(
                shuffledSideworks: currentState.shuffledSideworks,
                sideworks: currentState.sideworks,
                employees: currentState.employees,
                isSideworksEditShowing: currentState.isSideworksEditShowing,
                isEmployeesEditShowing: currentState.isEmployeesEditShowing,
                isSideworkEditShowing: false,
                isEmployeeEditShowing: currentState.isEmployeeEditShowing,
                newSideworkName: "",
                newEmployeeName: currentState.newEmployeeName,
                selectedSideworkID: "",
                selectedEmployeeID: currentState.selectedEmployeeID,
                employeeButtonText: currentState.employeeButtonText,
                sideworkButtonText: "Add",
                queue: currentState.queue
            )
            
        }else{
            self.state = self.state.doCopy(
                shuffledSideworks: currentState.shuffledSideworks,
                sideworks: currentState.sideworks,
                employees: currentState.employees,
                isSideworksEditShowing: currentState.isSideworksEditShowing,
                isEmployeesEditShowing: currentState.isEmployeesEditShowing,
                isSideworkEditShowing: true,
                isEmployeeEditShowing: currentState.isEmployeeEditShowing,
                newSideworkName: sidework.name,
                newEmployeeName: currentState.newEmployeeName,
                selectedSideworkID: sidework.id,
                selectedEmployeeID: currentState.selectedEmployeeID,
                employeeButtonText: currentState.employeeButtonText,
                sideworkButtonText: "Save",
                queue: currentState.queue
            )
            
        }
        
        currentState = (self.state.copy() as! AppState)
    }
    
    //MARK: - TOGGLE EDIT EMPLOYEE
    private func toggleEditEmployee(employee:Employee){
        
        var currentState = (self.state.copy() as! AppState)
        if(state.selectedEmployeeID == employee.id){
            self.state = self.state.doCopy(
                shuffledSideworks: currentState.shuffledSideworks,
                sideworks: currentState.sideworks,
                employees: currentState.employees,
                isSideworksEditShowing: currentState.isSideworksEditShowing,
                isEmployeesEditShowing: currentState.isEmployeesEditShowing,
                isSideworkEditShowing: currentState.isSideworkEditShowing,
                isEmployeeEditShowing: false,
                newSideworkName: currentState.newSideworkName,
                newEmployeeName: "",
                selectedSideworkID: currentState.selectedSideworkID,
                selectedEmployeeID: "",
                employeeButtonText: "Add",
                sideworkButtonText: currentState.sideworkButtonText,
                queue: currentState.queue
            )
            
        }else{
            self.state = self.state.doCopy(
                shuffledSideworks: currentState.shuffledSideworks,
                sideworks: currentState.sideworks,
                employees: currentState.employees,
                isSideworksEditShowing: currentState.isSideworksEditShowing,
                isEmployeesEditShowing: currentState.isEmployeesEditShowing,
                isSideworkEditShowing: currentState.isSideworkEditShowing,
                isEmployeeEditShowing: true,
                newSideworkName: currentState.newSideworkName,
                newEmployeeName: employee.name,
                selectedSideworkID: currentState.selectedSideworkID,
                selectedEmployeeID: employee.id,
                employeeButtonText: "Save",
                sideworkButtonText: currentState.sideworkButtonText,
                queue: currentState.queue
            )
            
        }
        
        currentState = (self.state.copy() as! AppState)
    }
    
    //MARK: - SET NEW SIDEWORK NAME
    private func setNewSideworkName(name : String){
        var currentState = (self.state.copy() as! AppState)
        self.state = self.state.doCopy(
            shuffledSideworks: currentState.shuffledSideworks,
            sideworks: currentState.sideworks,
            employees: currentState.employees,
            isSideworksEditShowing: currentState.isSideworksEditShowing,
            isEmployeesEditShowing: currentState.isEmployeesEditShowing,
            isSideworkEditShowing: currentState.isSideworkEditShowing,
            isEmployeeEditShowing: currentState.isEmployeeEditShowing,
            newSideworkName: name,
            newEmployeeName: currentState.newEmployeeName,
            selectedSideworkID: currentState.selectedSideworkID,
            selectedEmployeeID: currentState.selectedEmployeeID,
            employeeButtonText: currentState.employeeButtonText,
            sideworkButtonText: currentState.sideworkButtonText,
            queue: currentState.queue
        )
        
        currentState = (self.state.copy() as! AppState)
        
        
        
    }
    
    //MARK: - SET NEW EMPLOYEE NAME
    private func setNewEmployeeName(name : String){
        var currentState = (self.state.copy() as! AppState)
        self.state = self.state.doCopy(
            shuffledSideworks: currentState.shuffledSideworks,
            sideworks: currentState.sideworks,
            employees: currentState.employees,
            isSideworksEditShowing: currentState.isSideworksEditShowing,
            isEmployeesEditShowing: currentState.isEmployeesEditShowing,
            isSideworkEditShowing: currentState.isSideworkEditShowing,
            isEmployeeEditShowing: currentState.isEmployeeEditShowing,
            newSideworkName: currentState.newSideworkName,
            newEmployeeName: name,
            selectedSideworkID: currentState.selectedSideworkID,
            selectedEmployeeID: currentState.selectedEmployeeID,
            employeeButtonText: currentState.employeeButtonText,
            sideworkButtonText: currentState.sideworkButtonText,
            queue: currentState.queue
        )
        
        currentState = (self.state.copy() as! AppState)
        
        
        
    }
    
    //MARK: - DELETE SIDEWORK
    private func deleteSidework(sideworkID:String){
        deleteSideworkUseCase.execute(sideworkID: sideworkID)
            .collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                    if dataState != nil {
                        let data = dataState?.data
                        let message = dataState?.message
                        
                        
                        self.updateState()
                        
                        if(data != nil){
                            if (data as! [Sidework]).isEmpty {
                                print("sideworks are empty")
                            }
                            self.setSideworks(sideworks: data as! [Sidework])
                        }
                        
                        if message != nil{
                            self.handleMessageByUIComponentType(message!.build())
                        }
                    }
                })
    }
    
    //MARK: - DELETE EMPLOYEE
    private func deleteEmployee(employeeID:String){
        deleteEmployeeUseCase.execute(employeeID: employeeID)
            .collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                    if dataState != nil {
                        let data = dataState?.data
                        let message = dataState?.message
                        
                        
                        self.updateState()
                        
                        if(data != nil){
                            self.setEmployees(employees: data as! [Employee])
                        }
                        
                        if message != nil{
                            self.handleMessageByUIComponentType(message!.build())
                        }
                    }
                })
        
    }
    
    //MARK: - SAVE SIDEWORK
    private func saveSidework(sidework:Sidework){
        editSideworkUseCase.execute(sidework: sidework)
            .collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                    if dataState != nil {
                        let data = dataState?.data
                        let message = dataState?.message
                        
                        
                        self.updateState()
                        
                        if(data != nil){
                            self.setSideworks(sideworks: data as! [Sidework])
                        }
                        
                        if message != nil{
                            self.handleMessageByUIComponentType(message!.build())
                        }
                    }
                })
    }
    
    //MARK: - SAVE EMPLOYEE
    private func saveEmployee(employee:Employee){
        editEmployeeUseCase.execute(employee: employee)
            .collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                    if dataState != nil {
                        let data = dataState?.data
                        let message = dataState?.message
                        
                        
                        self.updateState()
                        
                        if(data != nil){
                            self.setEmployees(employees: data as! [Employee])
                        }
                        
                        if message != nil{
                            self.handleMessageByUIComponentType(message!.build())
                        }
                    }
                })
    }
    
    //MARK: - ASSIGN SIDEWORKS
    private func assignSideworks(){
        shuffleSideworksUseCase.execute()
            .collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                    if dataState != nil {
                        let data = dataState?.data
                        let message = dataState?.message
                        
                        
                        self.updateState()
                        
                        if(data != nil){
                            if (data as! [Sidework]).isEmpty {
                                print("Shuffled sideworks are empty")
                            }
                            self.setShuffledSideworks(sideworks: data as! [Sidework])
                        }
                        
                        if message != nil{
                            self.handleMessageByUIComponentType(message!.build())
                        }
                    }
                })
    }
    
    //MARK: - SET SHUFFLED SIDEWORKS
    private func setShuffledSideworks(sideworks:[Sidework]){
        var currentState = (self.state.copy() as! AppState)
        self.state = self.state.doCopy(
            shuffledSideworks: sideworks,
            sideworks: currentState.sideworks,
            employees: currentState.employees,
            isSideworksEditShowing: currentState.isSideworksEditShowing,
            isEmployeesEditShowing: currentState.isEmployeesEditShowing,
            isSideworkEditShowing: currentState.isSideworkEditShowing,
            isEmployeeEditShowing: currentState.isEmployeeEditShowing,
            newSideworkName: currentState.newSideworkName,
            newEmployeeName: currentState.newEmployeeName,
            selectedSideworkID: currentState.selectedSideworkID,
            selectedEmployeeID: currentState.selectedEmployeeID,
            employeeButtonText: currentState.employeeButtonText,
            sideworkButtonText: currentState.sideworkButtonText,
            queue: currentState.queue
        )
        
        currentState = (self.state.copy() as! AppState)
        
    }
    
    //MARK: - SET SIDEWORKS
    private func setSideworks(sideworks:[Sidework]){
        var currentState = (self.state.copy() as! AppState)
        self.state = self.state.doCopy(
            shuffledSideworks: currentState.shuffledSideworks,
            sideworks: sideworks,
            employees: currentState.employees,
            isSideworksEditShowing: currentState.isSideworksEditShowing,
            isEmployeesEditShowing: currentState.isEmployeesEditShowing,
            isSideworkEditShowing: false,
            isEmployeeEditShowing: currentState.isEmployeeEditShowing,
            newSideworkName: "",
            newEmployeeName: currentState.newEmployeeName,
            selectedSideworkID: currentState.selectedSideworkID,
            selectedEmployeeID: "",
            employeeButtonText: currentState.employeeButtonText,
            sideworkButtonText: "Add",
            queue: currentState.queue
        )
        
        currentState = (self.state.copy() as! AppState)
        
    }
    
    //MARK: - SET EMPLOYEES
    private func setEmployees(employees:[Employee]){
        var currentState = (self.state.copy() as! AppState)
        self.state = self.state.doCopy(
            shuffledSideworks: currentState.shuffledSideworks,
            sideworks: currentState.sideworks,
            employees: employees,
            isSideworksEditShowing: currentState.isSideworksEditShowing,
            isEmployeesEditShowing: currentState.isEmployeesEditShowing,
            isSideworkEditShowing: currentState.isSideworkEditShowing,
            isEmployeeEditShowing: false,
            newSideworkName: currentState.newSideworkName,
            newEmployeeName: "",
            selectedSideworkID: currentState.selectedSideworkID,
            selectedEmployeeID: "",
            employeeButtonText: "Add",
            sideworkButtonText: currentState.sideworkButtonText,
            queue: currentState.queue
        )
        currentState = (self.state.copy() as! AppState)
        
    }
    
    //MARK: - SHOW MAIN SCREEN
    private func showMainScreen(sideworks:[Sidework]){
        
        var currentState = (self.state.copy() as! AppState)
        self.state = self.state.doCopy(
            shuffledSideworks: currentState.shuffledSideworks,
            sideworks: currentState.sideworks,
            employees: currentState.employees,
            isSideworksEditShowing: false,
            isEmployeesEditShowing: false,
            isSideworkEditShowing: currentState.isSideworkEditShowing,
            isEmployeeEditShowing: currentState.isEmployeeEditShowing,
            newSideworkName: currentState.newSideworkName,
            newEmployeeName: currentState.newEmployeeName,
            selectedSideworkID: currentState.selectedSideworkID,
            selectedEmployeeID: currentState.selectedEmployeeID,
            employeeButtonText: currentState.employeeButtonText,
            sideworkButtonText: currentState.sideworkButtonText,
            queue: currentState.queue
        )
        
        currentState = (self.state.copy() as! AppState)
        
    }
    
    //MARK: - SHOW EDIT EMPLOYEES
    private func showEditEmployees(){
        
        var currentState = (self.state.copy() as! AppState)
        self.state = self.state.doCopy(
            shuffledSideworks: currentState.shuffledSideworks,
            sideworks: currentState.sideworks,
            employees: currentState.employees,
            isSideworksEditShowing: currentState.isSideworksEditShowing,
            isEmployeesEditShowing: !currentState.isEmployeesEditShowing,
            isSideworkEditShowing: currentState.isSideworkEditShowing,
            isEmployeeEditShowing: currentState.isEmployeeEditShowing,
            newSideworkName: currentState.newSideworkName,
            newEmployeeName: currentState.newEmployeeName,
            selectedSideworkID: currentState.selectedSideworkID,
            selectedEmployeeID: currentState.selectedEmployeeID,
            employeeButtonText: currentState.employeeButtonText,
            sideworkButtonText: currentState.sideworkButtonText,
            queue: currentState.queue
        )
        
        currentState = (self.state.copy() as! AppState)
        
    }
    
    //MARK: - SHOW EDIT SIDEWORKS
    private func showEditSideworks(){
        
        var currentState = (self.state.copy() as! AppState)
        self.state = self.state.doCopy(
            shuffledSideworks: currentState.shuffledSideworks,
            sideworks: currentState.sideworks,
            employees: currentState.employees,
            isSideworksEditShowing: !currentState.isSideworksEditShowing,
            isEmployeesEditShowing: currentState.isEmployeesEditShowing,
            isSideworkEditShowing: currentState.isSideworkEditShowing,
            isEmployeeEditShowing: currentState.isEmployeeEditShowing,
            newSideworkName: currentState.newSideworkName,
            newEmployeeName: currentState.newEmployeeName,
            selectedSideworkID: currentState.selectedSideworkID,
            selectedEmployeeID: currentState.selectedEmployeeID,
            employeeButtonText: currentState.employeeButtonText,
            sideworkButtonText: currentState.sideworkButtonText,
            queue: currentState.queue
        )
        
        currentState = (self.state.copy() as! AppState)
    }
    
    //MARK: - ERROR HANDLER
    private func handleMessageByUIComponentType(_ message:GenericMessageInfo){
        switch message.uiComponentType {
        case UIComponentType.Dialog():
            appendToQueue(message: message)
        case UIComponentType.None():
            print("\(message.description)")
        default:
            doNothing()
        }
    }
    
    func doNothing(){
        
    }
    
    
    func shouldShowDialog(){
        let currentState = self.state.copy() as! AppState
        showDialog = currentState.queue.items.count > 0
    }
    
    func appendToQueue(message: GenericMessageInfo){
        let currentState = self.state.copy() as! AppState
        let queue = currentState.queue
        let queueUtil = GenericMessageInfoQueueUtil()
        
        if(!queueUtil.doesMessageAlreadyExistInQueue(queue: queue, messageInfo: message)){
            queue.add(element: message)
            updateState(queue: queue)
        }
    }
    
    func removeHeadFromQueue(){
        let currentState = self.state.copy() as! AppState
        let queue = currentState.queue
        do {
            try queue.remove()
            updateState(queue: queue)
            
        }
        catch{
            print("\(error)")
        }
    }
}
